### PR TITLE
TUIActivityIndicator Revamp!

### PR DESCRIPTION
**Pull request by avaidyam**
The next control on my hitlist for revamping: `TUIActivityIndicatorView`! I've renamed it to `TUIActivityIndicator` to remove the explanatory `View` from its name, and made it match the OS X/iOS activity indicator with customizable animations.

### CODE REVIEW

**Comment by jspahrsummers** [comment]
:-1: This is impressive work, but inappropriate for TwUI proper. `TUIActivityIndicatorView` has near parity with `UIActivityIndicatorView` right now – any extension of the scope beyond that is more appropriate for a third-party class.

**Comment by avaidyam** [comment]
I'd like to mention that `TUIActivityIndicatorView` looks and acts nothing like `UIActivityIndicatorView`...

**Comment by jspahrsummers** [comment]
The API has near parity, which is what I was referring to. If the appearance is different, that may have been a deliberate design choice. In any case, appearance changes are something we can discuss in issues or PRs – just because it doesn't look the same doesn't mean it should be replaced with a completely new API.

**Comment by jspahrsummers** [comment]
Since it seems like you have lots of ideas like this, why not start a project for reusable TwUI views and controls? You could put these sorts of things into a library like that without bumping up against the scope of the TwUI framework itself.

**Comment by jwilling** [comment]
One thing to note is that this might've been easier to implement if you had used a `CAReplicatorLayer` for the gears. :smiley:

**Comment by avaidyam** [comment]
But it was already easy to implement? I'll look into `CAReplicatorLayer` though, thanks!

**Comment by avaidyam** [comment]
As it is right now, I don't think TwUI does much. It has a half baked text engine, a powerful layer-backed view system, but nothing that works out of the box. With AppKit, it takes just a few lines to get a good app design working. With TwUI it takes MANY subclasses, and a lot of research and bug-fixing.  

I do understand that you don't want bloat added, but quoting the original "unveiling" post by Twitter, "And, since TwUI isn’t bound by the constraints of an existing API, developers can experiment with new features like block-based drawRect and layout.  …  We’ll continue to develop additional features and make improvements. And, we encourage you to experiment, as that will help us build a robust and exciting UI framework for the Mac."

And that's exactly what I'm doing, experimenting and developing new features for TwUI. I'd feel like as long as it's not something absolutely insane (like, for example, remote `CALayer` rendering and event handling for VNC-like uses), it should be fine to add in.

Back to the scope of this PR, quoting the `README`, "2. It's something so basic or important that TwUI really should have it." A `TUIActivityIndicatorView` is something I'd consider quite basic, and _preferably_ one that works as advertised. Following the first quote, the additional customization features should be perfectly okay too. Looking at the code, all I did to add these extra features was expose methods that would have been otherwise private, and add a set of animations and an indicator frame to _restore_ the old `TUIActivityIndicatorView` look. It's a simple hot swap of sorts. That's why I believe this PR should be okay to merge.

Sorry if I've come across as rude or raging, I'm just quite frustrated that the work I'm putting into a public framework repository is being turned down for being "too much". I'd much rather share these additions in the main repository so it becomes easier to use with a single download, out of the box, than add it to another repository as a subclass.

**Comment by joshaber** [comment]
Our philosophy is that TwUI should be a foundation for UI. It shouldn't provide everything under the sun—just the core that anyone would need to make an app.

I'd say this situation is very similar to one in UIKit. You can find probably 100s of open source, customized UIKit activity indicators on the internet. But none of them belong in UIKit proper, even if UIKit were open source. `UIActivityIndicator` solves the general case simply.

It's impossible to overestimate the value of simplicity in an API. When it's a choice between simple API and complex API, we'll probably always side with simplicity.

Like @jspahrsummers, this is cool and you should certainly put it somewhere. But TwUI isn't that place.

To put it another way, we'd rather TwUI were a core around which people built cool extensions than a monolithic, all-encompassing solution.

**Comment by avaidyam** [comment]
Thank you for that, it really helps me understand what TwUI _is to become_. On that note, would this PR be up for consideration again, if the excess features were removed? At its minimum, I wanted the activity indicator to look and act like iOS and OS X. It was when I realized that the animations could be replaced with a simple array, and the drawing replaced with a drawRect block to the inner knob, that I added these properties. To be honest, all I did was factor out the usual mechanism into public methods, which is why I'm heavily contesting the closing of this PR. I can, and _will_ make the API simpler, if that would let this PR go through.

File: lib/UIKit/TUIActivityIndicator.h [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+// The basic appearance of the activity indicator.
+// The default value is TUIActivityIndicatorStyleWhite.
+@property (nonatomic, assign) TUIActivityIndicatorStyle activityIndicatorStyle;
+
+// The layerProxy allows you to modify layer properties on the indicator
+// as it rotates as well, while using the basic .layer property allows you
+// to modify stationary layer properties - those that are not affected by
+// the animations on the indicator.
+@property (nonatomic, readonly) CALayer *layerProxy;
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: The layer proxy is just a one-line accessor to the internal `indicator.layer`. It just lets you do fancy things.

File: lib/UIKit/TUIActivityIndicator.h [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+// the indicator layer's hidden property to YES when is not animating.
+// If the hidesWhenStopped property is NO, the indicator layer is not hidden
+// when animation stops.
+@property (nonatomic, assign) BOOL hidesWhenStopped;
+
+// Returns YES if the indicator is currently animating, otherwise NO.
+@property (nonatomic, readonly, getter = isAnimating) BOOL animating;
+
+// Changes the speed of the indicator animation. Defaults to 1.0 seconds.
+@property (nonatomic, assign) CGFloat animationSpeed;
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: The animationSpeed is another single-line property that lets you speed up or down the animations.

File: lib/UIKit/TUIActivityIndicator.h [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+// Sets the animations that the indicator will perform while animating.
+// Any duration, timing, fill mode, or repeat count information will be
+// discarded to ensure consistent indicator animations. Due to the
+// mutable nature of the animations array, you may combine several
+// unrelated animations to achieve a single complex animation. Also note
+// that anything that is not a CAAnimation object will be discarded.
+@property (nonatomic, strong) NSMutableArray *animations;
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: As I was factoring out the animation to become more iOS/OS X-like, I realized that it was already an array, so why not make this array public?

File: lib/UIKit/TUIActivityIndicator.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+// The indicator frame can be used to apply custom indicator drawing,
+// and a combination of the indicatorFrame and the animations array,
+// a completely custom look can be applied. This will only apply if
+// the activityIndicatorStyle is set to TUIActivityIndicatorStyleCustom.
+@property (nonatomic, copy) TUIViewDrawRect indicatorFrame;
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: A simple 3 line setter/getter to the internal `indicator.drawRect`.

File: lib/UIKit/TUIActivityIndicator.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+// Restarts the animation of the indicator quietly, without triggering an
+// animation start or stop. This method is best used when drawing or
+// animation code is updated.
+- (void)refreshAnimations;
+
+@end
+
+// Returns an indicator frame with a basic centered gray circle.
+extern TUIViewDrawRect TUIActivityIndicatorCircleFrame();
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: Functions the same way as `TUITextView`'s `drawFrame` property, and I thought it would be okay, since the idea already existed in TwUI.

File: lib/UIKit/TUIActivityIndicator.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+- (void)startAnimating;
+
+// Stops the animation of the indicator. When animation is stopped, the
+// indicator is hidden if hidesWhenStopped is YES.
+- (void)stopAnimating;
+
+// Restarts the animation of the indicator quietly, without triggering an
+// animation start or stop. This method is best used when drawing or
+// animation code is updated.
+- (void)refreshAnimations;
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: Again, as I was factoring out methods, I decided to make one or two public.

File: lib/UIKit/TUIActivityIndicator.h [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+// The default is 12 teeth with a 2.0 pixel width. The color is decided
+// by the preset activityIndicatorStyle.
+extern TUIViewDrawRect TUIActivityIndicatorGearFrame(CGFloat toothCount,
+													 CGFloat toothWidth,
+													 NSColor *toothColor);
+
+// Returns an array of animations designed to pulse the indicator. Consists
+// of a scaling transformation paired with an opacity blend for pulsing.
+// Pass a CGFloat for the peak opacity of the pulse animation.
+extern NSArray * TUIActivityIndicatorPulseAnimations(CGFloat);
```
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: Similar to above, but instead of pre-packaged `drawFrames`, it's a pre-packaged animation array. (It's usually a single animation, too).

File: lib/UIKit/TUIView+Layout.h [hunk 8]
```diff
@@ -25,4 +25,6 @@
 
 - (TUIView *)relativeViewForName:(NSString *)name;
 
+- (void)setNeedsUpdateConstraints;
```
Comment (avaidyam) [inline lib/UIKit/TUIView+Layout.h]: Mirrors `UIView` and `NSView`. Now the `TUILayoutManager` can go fully private. Another PR for that later, though.

**Comment by avaidyam** [comment]
@jspahrsummers @jwilling @joshaber I've highlighted the changes (which I think are minuscule) I've made to the `TUIActivityIndicator`. I can remove any of these, if you'd like.

File: lib/UIKit/TUIActivityIndicator.h (hunk 3 above)
Comment (jwilling) [inline lib/UIKit/TUIActivityIndicator.h]: I don't think making a mutable array public is a good idea. Since you're only accepting `CAAnimations`, why not just add a method that adds a single `CAAnimation` object onto the private array of animations?

Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicator.h]: I could do that….. :D

**Comment by jspahrsummers** [comment]
As per @joshaber and my comments, this will only be considered if the API matches `UIActivityIndicatorView`. We're just talking in circles now.

**Comment by avaidyam** [comment]
@jspahrsummers I've made the API clear and refactored. The API _does_ match `UIActivityIndicatorView`, with the exception of `layerProxy` and `indicatorFrame` which are both literally one-line accessors into the underlying indicator view, which `UIActivityIndicatorView` explicitly hides from you.

@jwilling I've moved the animations array into 2-line methods. :)

**Comment by jwilling** [comment]
@galaxas0 :flushed: That might be worse than it was before. If you want to keep the ability to add an array of animations, it's probably good just to make it a immutable array. Adding a mutable property of this type just opens doors for all sorts of open ends about behaviors adding animations after they've begun.

**Comment by avaidyam** [comment]
I have provisions for that-- they call `-refreshAnimations` when added.

**Comment by avaidyam** [comment]
Wait-- just kidding, I see what you mean. +_____+

**Comment by avaidyam** [comment]
I've instead reverted it to a mutable array, but made the note to call the refresh method. I've made provisions for any sort of refreshing issues already, as it's inherent.

**Comment by jwilling** [comment]
I just don't understand your reasoning for making the array readonly and mutable. If you don't want them to modify the array during animation, then make it readwrite and immutable. Better not presume your users will follow advice in a comment.

**Comment by avaidyam** [comment]
Oh, you _can_ modify the array during animation. That's why I did that. Just add a call to `refreshAnimations`. The array won't be affected either way during animations because I've offloaded the animations into a `CAAnimationGroup`.

**Comment by avaidyam** [comment]
@jspahrsummers I've removed the extraneous features. Is there anything else you'd like me to remove?

**Comment by jspahrsummers** [comment]
I'll review this if you can pare down the interface to:

``` objc
typedef enum {
    TUIActivityIndicatorViewStyleWhite,
    TUIActivityIndicatorViewStyleGray
} TUIActivityIndicatorViewStyle;

@interface TUIActivityIndicatorView : TUIView

@property (nonatomic, assign) TUIActivityIndicatorViewStyle activityIndicatorViewStyle;
@property (nonatomic, assign) BOOL hidesWhenStopped;
@property (nonatomic, readonly, getter = isAnimating) BOOL animating;

- (id)initWithActivityIndicatorViewStyle:(TUIActivityIndicatorStyle)style;
- (void)startAnimating;
- (void)stopAnimating;

@end
```

Note the name as well. I think we should follow the naming established by `UIActivityIndicatorView`, rather than `NSProgressIndicator`, since the former is more recent.

**Comment by avaidyam** [comment]
I was going with `TUIActivityIndicator` since the addition of `View` is redundant, because it's a `TUIView` subclass by default. And why not the `animationSpeed`?

**Comment by jspahrsummers** [comment]
It's not automatically implied that something within TwUI is a subclass of `TUIView`. There are several subclasses of AppKit classes within the framework. And, again, I'm talking about UIKit parity on this point.

> And why not the animationSpeed?

There should be no reason to customize the speed of the animation if the animation itself is not customizable.

**Comment by avaidyam** [comment]
But I did make the animation customizable? Would it be alright to remove this `animationSpeed` and let animations have their own times?

**Comment by jspahrsummers** [comment]
The interface I proposed above intentionally does not include customizable animations. The feature is unnecessary and difficult to use (not to mention maintain).

Another thing to keep in mind: TwUI is open source. If someone doesn't like how a component works, or wants more customization, they can always build their own based on the existing source code. We don't need to provide the kitchen sink.

**Comment by avaidyam** [comment]
I've stripped the interface down a fair amount, and I am ready to remove more, but I'm really concerned that there's no way to add custom animations anymore. Perhaps two methods like `beginUpdates` and `endUpdates` for `UITableView` would work for this. I'm very _very_ keen on the animations being easy to modify, but I completely understand that just putting it out there is _not_ what TwUI wants to do. Would you be open to any other methods of "inserting" animations?

**Comment by jspahrsummers** [comment]
Custom animations do not have to be part of `TUIActivityIndicatorView`. The class is meant to be very basic.

If someone wants a different animation, it's trivial to write a new class to do it. Built-in TwUI classes have no advantage over what a user of the framework could write themselves, so we can't use "it's hard to do" as a reason to put more stuff into TwUI.

**Comment by avaidyam** [comment]
@jspahrsummers I understand, and I've removed the customization features. The header is now as simple as you've described.

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox_with_mail:

File: lib/UIKit/TUIActivityIndicator.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+@property (nonatomic, readonly, getter = isAnimating) BOOL animating;
+
+// Initializes the activity indicator with the style of the indicator.
+// You can set and retrieve the style of a activity indicator through
+// the activityIndicatorViewStyle property. See TUIActivityIndicatorStyle
+// for descriptions of the style constants. 
+- (id)initWithActivityIndicatorStyle:(TUIActivityIndicatorStyle)style;
+
+// Initializes the activity indicator with both style and frame.
+- (id)initWithFrame:(CGRect)frame andActivityIndicatorStyle:(TUIActivityIndicatorStyle)style;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicator.h]: Parameter names should never include `and` in Cocoa.

File: lib/UIKit/TUIActivityIndicator.m [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+#import "TUILayoutConstraint.h"
+#import "TUICGAdditions.h"
+
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect){0, 0, 32, 32};
+static TUIActivityIndicatorStyle TUIActivityIndicatorDefaultStyle = TUIActivityIndicatorStyleWhite;
+
+static CGFloat const TUIActivityIndicatorDefaultToothCount = 12.0f;
+static CGFloat const TUIActivityIndicatorDefaultToothWidth = 2.0f;
+
+static TUIViewDrawRect TUIActivityIndicatorGearFrame(CGFloat toothCount, CGFloat toothWidth, NSColor *toothColor) {
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicator.m]: Why is this still here?

File: lib/UIKit/TUIActivityIndicator.m [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+			[[toothColor colorWithAlphaComponent:alpha] setFill];
+			
+			CGContextRotateCTM(ctx, 1 / toothCount * (M_PI * 2.0f));
+			CGRect toothRect = CGRectMake(-toothWidth / 2.0f, -radius, toothWidth, ceilf(radius * 0.54f));
+			CGContextFillRoundRect(ctx, toothRect, toothWidth / 2.0f);
+		}
+	} copy];
+}
+
+static NSArray * TUIActivityIndicatorGearAnimations(CGFloat frameCount) {
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicator.m]: Why is this still here?

File: lib/UIKit/TUIActivityIndicator.h [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+typedef enum {
+	
+	// The standard white style of indicator.
+	TUIActivityIndicatorStyleWhite,
+	
+	// The standard gray style of indicator.
+	TUIActivityIndicatorStyleGray
+} TUIActivityIndicatorStyle;
+
+@interface TUIActivityIndicator : TUIView
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicator.h]: This class still needs to be renamed to `TUIActivityIndicatorView`.

**Comment by jspahrsummers** [comment]
This code has changed very little – you've only modified the public interface to hide the features that we declined to add.

Please reopen this once you've _actually_ reduced the scope.

**Comment by avaidyam** [comment]
Sure, I'll fix that.

**Comment by avaidyam** [comment]
@jspahrsummers I've reduced the scope and removed the reusability blocks. If there's anything I missed, I can remove that as well.

**Comment by avaidyam** [comment]
@jspahrsummers Is there anything else wrong with this that I should change? It's been 4 days since I've updated it.

**Comment by jspahrsummers** [comment]
Please reopen this pull request if you think it's ready for review.

**Comment by avaidyam** [comment]
I'm not given the option to re-open it on github.com as there's only a "Comment" button below.

**Comment by jwilling** [comment]
Odd. I'll open it for you since you think it's ready.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 13]
```diff
@@ -15,80 +15,143 @@
  */
 
 #import "TUIActivityIndicatorView.h"
+#import "TUILayoutConstraint.h"
+#import "TUICGAdditions.h"
+
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect){0, 0, 32, 32};
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: Initializing structs based on member order is fragile. Please use `CGRectMake` or C's designated initializer syntax.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 14]
```diff
@@ -15,80 +15,143 @@
...
+static TUIActivityIndicatorStyle TUIActivityIndicatorDefaultStyle = TUIActivityIndicatorStyleWhite;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: This should be `const`.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 15]
```diff
@@ -15,80 +15,143 @@
...
+	return [self initWithFrame:frame activityIndicatorStyle:TUIActivityIndicatorDefaultStyle];
+}
+
+- (id)initWithActivityIndicatorStyle:(TUIActivityIndicatorStyle)style {
+	return [self initWithFrame:TUIActivityIndicatorDefaultFrame activityIndicatorStyle:style];
+}
+
+- (void)setHidesWhenStopped:(BOOL)hide {
+	_hidesWhenStopped = hide;
+	if(!self.animating && !hide && self.proxyIndicator.hidden)
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: Checking `self.proxyIndicator.hidden` is unnecessary here.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 16]
```diff
@@ -15,80 +15,143 @@
...
+		self.proxyIndicator.hidden = NO;
+	else
+		self.proxyIndicator.hidden = YES;
+}
+
+- (void)setActivityIndicatorStyle:(TUIActivityIndicatorStyle)style {
+	_activityIndicatorStyle = style;
+	
+	self.proxyIndicator.drawRect = ^(TUIView *indicator, CGRect rect) {
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: For clarity, this block should be set on initialization (not when the style is changed), and can read from `weakSelf.activityIndicatorStyle` instead of closing over the argument.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 17]
```diff
@@ -15,80 +15,143 @@
...
+		NSColor *toothColor = [NSColor whiteColor];
+		
+		if(style == TUIActivityIndicatorStyleGray)
+			toothColor = [NSColor grayColor];
+		else if(style == TUIActivityIndicatorStyleBlack)
+			toothColor = [NSColor blackColor];
+		
+		CGContextRef ctx = TUIGraphicsGetCurrentContext();
+		CGContextTranslateCTM(ctx, radius, radius);
+		CGContextClearRect(ctx, rect);
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: This is done by `clearsContextBeforeDrawing`, and should be unnecessary here.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 18]
```diff
@@ -15,80 +15,143 @@
...
+		}
+	};
+	
+	[self refreshAnimations];
+}
+
+- (void)startAnimating {
+	if(!self.animating) {
+		self.proxyIndicator.hidden = NO;
+		_animating = YES;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: This should use a setter. Avoid setting ivars directly whenever possible.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 19]
```diff
@@ -15,80 +15,143 @@
...
+		NSMutableArray *times = [NSMutableArray array];
+		for(int i = 0; i < TUIActivityIndicatorDefaultToothCount + 1; i++) {
+			[times addObject:@(1.0 * (i / TUIActivityIndicatorDefaultToothCount))];
 		}
 		
-		_animating = YES;
+		CAKeyframeAnimation *rotate = [CAKeyframeAnimation animationWithKeyPath:@"transform.rotation.z"];
+		rotate.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear];
+		rotate.calculationMode = kCAAnimationDiscrete;
+		rotate.repeatCount = INT_MAX;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: This should be `HUGE_VALF`, per the [CAMediaTiming documentation](https://developer.apple.com/library/mac/documentation/GraphicsImaging/Reference/CAMediaTiming_protocol/Introduction/Introduction.html#//apple_ref/doc/uid/TP40004511-CH1-SW5).

File: lib/UIKit/TUIView+Layout.h (hunk 8 above)
Comment (jspahrsummers) [inline lib/UIKit/TUIView+Layout.h]: This shouldn't be in this pull request.

**Comment by jspahrsummers** [comment]
:mailbox_with_mail:

File: lib/UIKit/TUIActivityIndicatorView.m (hunk 13 above)
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicatorView.m]: Actually, LLVM 4.2 provides a rather nifty way to do this by defining the struct member like so:

```
(CGRect) {
    .origin.x = 0, 
    .origin.y = 0,
    .size.width = 32, 
    .size.height = 32
};
```

Comment (jwilling) [inline lib/UIKit/TUIActivityIndicatorView.m]: That's not new in LLVM. :wink: Was introduced in C99 if I recall correctly.

Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicatorView.m]: Wow, really? I read it in one of the Clang release notes.... or am I confusing things... +____+

File: lib/UIKit/TUIActivityIndicatorView.m (hunk 19 above)
Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicatorView.m]: Nice catch-- didn't know that. Time to read up on that!

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox:

File: lib/UIKit/TUIActivityIndicatorView.m (hunk 13 above)
Comment (jwilling) [inline lib/UIKit/TUIActivityIndicatorView.m]: And as a matter of fact, I believe that could be shortened to the following: 

```
CGRect rect = (CGRect) {
    .size.width = 32, 
    .size.height = 32
};
```

as the origin will be initialized to zero point automatically if not provided. Anyway, it's irrelevant. :stuck_out_tongue_winking_eye:

Comment (avaidyam) [inline lib/UIKit/TUIActivityIndicatorView.m]: I guess it's better to be explicit? ;D :dart:

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 20]
```diff
@@ -15,80 +15,149 @@
  */
 
 #import "TUIActivityIndicatorView.h"
+#import "TUILayoutConstraint.h"
+#import "TUICGAdditions.h"
+
+static TUIActivityIndicatorStyle const TUIActivityIndicatorDefaultStyle = TUIActivityIndicatorStyleWhite;
+static CGRect const TUIActivityIndicatorDefaultFrame = (CGRect) {
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: `(CGRect)` is unnecessary here. The type should be inferred from the variable being assigned into.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 21]
```diff
@@ -15,80 +15,149 @@
...
+	.origin.x = 0,
+	.origin.y = 0,
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: These can be condensed into `.origin = CGPointZero`, or omitted entirely (as noted by @jwilling).

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 22]
```diff
@@ -15,80 +15,149 @@
...
+	.size.width = 32,
+	.size.height = 32
+};
+
+static CGFloat const TUIActivityIndicatorDefaultToothCount = 12.0f;
+static CGFloat const TUIActivityIndicatorDefaultToothWidth = 2.0f;
+
+@interface TUIActivityIndicatorView ()
+
+@property (nonatomic, strong) TUIView *proxyIndicator;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: This should be `readonly`, as it is only ever set once.

File: lib/UIKit/TUIActivityIndicatorView.m [hunk 23]
```diff
@@ -15,80 +15,149 @@
...
+		
+		self.activityIndicatorStyle = style;
+		self.hidesWhenStopped = YES;
 		
-		spinner = [[TUIView alloc] initWithFrame:self.bounds];
-		spinner.backgroundColor = [NSColor blackColor];
-		spinner.alpha = 0.2;
-		spinner.layer.cornerRadius = 10.0;
-		[self addSubview:spinner];
+		__block TUIActivityIndicatorView *weakSelf = self;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.m]: `__block` does not make a variable weak under ARC. This should be:

``` objc
__unsafe_unretained TUIActivityIndicatorView *weakSelf = self;
```

File: lib/UIKit/TUIView+Layout.h (hunk 8 above)
Comment (jspahrsummers) [inline lib/UIKit/TUIView+Layout.h]: ^

**Comment by jspahrsummers** [comment]
This makes the activity indicator animate in the wrong direction. It animates counterclockwise (making it look like progress is being rewound), when it should animate clockwise.

**Comment by avaidyam** [comment]
@jspahrsummers I've made some style fixes and it now animates properly. :mailbox_with_mail:

File: lib/UIKit/TUIActivityIndicatorView.h [hunk 24]
```diff
@@ -17,23 +17,48 @@
...
-	TUIActivityIndicatorViewStyleGray,
-} TUIActivityIndicatorViewStyle;
+	
+	// The standard white style of indicator.
+	TUIActivityIndicatorStyleWhite,
+	
+	// The standard gray style of indicator.
+	TUIActivityIndicatorStyleGray,
+	
+	// The standard gray style of indicator.
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.h]: Incorrect comment here.

File: lib/UIKit/TUIActivityIndicatorView.h [hunk 25]
```diff
@@ -17,23 +17,48 @@
...
-{
-    BOOL                           _animating;
-    TUIActivityIndicatorViewStyle _activityIndicatorViewStyle;
-	
-	TUIView *spinner;
-}
 
-- (id)initWithActivityIndicatorStyle:(TUIActivityIndicatorViewStyle)style;
+// The basic appearance of the activity indicator.
+// The default value is TUIActivityIndicatorStyleWhite.
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.h]: We should use the gray style as the default, like it was before.

File: lib/UIKit/TUIActivityIndicatorView.h [hunk 26]
```diff
@@ -17,23 +17,48 @@
 #import "TUIView.h"
 
 typedef enum {
-	TUIActivityIndicatorViewStyleGray,
-} TUIActivityIndicatorViewStyle;
+	
+	// The standard white style of indicator.
+	TUIActivityIndicatorStyleWhite,
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.h]: The values in this enum should be prefixed with `TUIActivityIndicatorViewStyle`.

File: lib/UIKit/TUIActivityIndicatorView.h [hunk 27]
```diff
@@ -17,23 +17,48 @@
...
+	
+	// The standard gray style of indicator.
+	TUIActivityIndicatorStyleGray,
+	
+	// The standard black style of indicator.
+	TUIActivityIndicatorStyleBlack
+} TUIActivityIndicatorStyle;
```
Comment (jspahrsummers) [inline lib/UIKit/TUIActivityIndicatorView.h]: This enum should be `TUIActivityIndicatorViewStyle`.

File: lib/UIKit/TUIView+Layout.h (hunk 8 above)
Comment (jspahrsummers) [inline lib/UIKit/TUIView+Layout.h]: ^

**Comment by jspahrsummers** [comment]
Added some more notes. Also commented again on the `setNeedsUpdateConstraints` comment that you've missed a few times now.

**Comment by avaidyam** [comment]
@jspahrsummers Ah, my bad. It should be _really_ fixed now... college apps drive me nuts.

**Comment by avaidyam** [comment]
@jspahrsummers :mailbox_with_mail:

**Comment by jspahrsummers** [comment]
Thanks!

**Comment by avaidyam** [comment]
At last! :D

**Comment by jwilling** [comment]
I'm not altogether sure this is without bugs. This was taken from the scrolling example:

**Comment by avaidyam** [comment]
Whoah, uhh, that's weird and totally not expected. That didn't happen before either…. I'll put in a PR to fix this weirdness.

**Comment by avaidyam** [comment]
Wait, this is absolutely confusing... how is this happening.....????
